cpp/src/parquet/encoding_test.cc

```diff
@@ -862,5 +862,215 @@ TEST_F(DictEncoding, CheckDecodeIndicesNoNulls) {
+  GenerateData<T>(nvalues, data.data(), NULLPTR);
+  std::unique_ptr<TypedEncoder<DType>> encoder =
+      MakeTypedEncoder<DType>(Encoding::BYTE_STREAM_SPLIT);
+  encoder->Put(data.data(), static_cast<int>(data.size()));
+  std::shared_ptr<Buffer> buffer = encoder->FlushValues();
+
+  std::unique_ptr<TypedDecoder<DType>> decoder =
+      MakeTypedDecoder<DType>(Encoding::BYTE_STREAM_SPLIT);
+  decoder->SetData(static_cast<int>(data.size()), buffer->mutable_data(),
+                   static_cast<int>(buffer->size()));
+  std::vector<T> decoded_data(nvalues);
+  int num_decoded_elements = decoder->Decode(decoded_data.data(), nvalues);
+  ASSERT_EQ(nvalues, num_decoded_elements);
+  for (size_t i = 0U; i < decoded_data.size(); ++i) {
+    ASSERT_EQ(data[i], decoded_data[i]);
+  }
+}
+}  // namespace
+
+// Check that the encoder can handle empty input.
+TEST(ByteStreamSplitEncodeDecode, EncodeZeroLenInput) {
+  std::unique_ptr<TypedEncoder<FloatType>> encoder =
+      MakeTypedEncoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  encoder->Put(NULL, 0);
+  ASSERT_EQ(0U, encoder->EstimatedDataEncodedSize());
+  std::shared_ptr<Buffer> encoded_buffer = encoder->FlushValues();
+  ASSERT_EQ(0, encoded_buffer->size());
+}
+
+template <typename DType>
+bool TestCreatingByteStreamSplitEncoder() {
+  try {
+    MakeTypedEncoder<DType>(Encoding::BYTE_STREAM_SPLIT);
+    return true;
+  } catch (ParquetException&) {
+    return false;
+  }
+}
+
+template <typename DType>
+bool TestCreatingByteStreamSplitDecoder() {
+  try {
+    MakeTypedDecoder<DType>(Encoding::BYTE_STREAM_SPLIT);
+    return true;
+  } catch (ParquetException&) {
+    return false;
+  }
+}
+
+// Check that the encoder can handle input with one element.
+TEST(ByteStreamSplitEncodeDecode, EncodeOneLenInput) {
+  std::unique_ptr<TypedEncoder<FloatType>> encoder =
+      MakeTypedEncoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+
+  const float value = 1.0f;
+  encoder->Put(&value, 1);
+
+  const int64_t estimated_num_bytes = encoder->EstimatedDataEncodedSize();
+  ASSERT_EQ(4, estimated_num_bytes);
+
+  std::shared_ptr<Buffer> encoded_buffer = encoder->FlushValues();
+  ASSERT_EQ(4U, encoded_buffer->size());
+  const uint8_t* mutable_data = encoded_buffer->mutable_data();
+
+  const uint8_t* value_as_bytes = reinterpret_cast<const uint8_t*>(&value);
+  ASSERT_EQ(value_as_bytes[0], mutable_data[0]);
+  ASSERT_EQ(value_as_bytes[1], mutable_data[1]);
+  ASSERT_EQ(value_as_bytes[2], mutable_data[2]);
+  ASSERT_EQ(value_as_bytes[3], mutable_data[3]);
+
+  ASSERT_EQ(0, encoder->EstimatedDataEncodedSize());
+
+  encoded_buffer = encoder->FlushValues();
+  ASSERT_EQ(0, encoded_buffer->size());
+}
+
+// Check that the decoder can handle empty input.
+TEST(ByteStreamSplitEncodeDecode, DecodeZeroLenInput) {
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  decoder->SetData(0, NULL, 0);
+  ASSERT_EQ(0U, decoder->Decode(NULL, 0));
```

Comment:
Similar question re: code reuse

Reply:
Since it's a refactoring for all encoder, I can do this in a follow-up patch.